File: app/src/main/java/com/techatonic/sudokuapp/backend/sudokutypes/KillerSudokuType.java
package com.techatonic.sudokuapp.backend.sudokutypes;

import androidx.core.util.Pair;

import java.util.ArrayList;
import java.util.List;

public class KillerSudokuType extends ClassicSudokuType{
    private ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages = new ArrayList<>();

    public KillerSudokuType(SudokuType type, int[][] grid) {
        super(type, grid);
    }

    public KillerSudokuType() {
        super(SudokuType.Killer);
    }

    public KillerSudokuType(ClassicSudokuType sudoku) {
        super(sudoku);
    }

    public KillerSudokuType(int[][] grid) {
        super(grid);
    }

    public KillerSudokuType(SudokuType type, int[][] grid, ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages) {
        super(type, grid);
        this.cages = cages;
    }

    public KillerSudokuType(SudokuType type, ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages) {
        super(type);
        this.cages = cages;
    }

    public KillerSudokuType(ClassicSudokuType sudoku, ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages) {
        super(sudoku);
        this.cages = cages;
    }

    public KillerSudokuType(int[][] grid, ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages) {
        super(grid);
        this.cages = cages;
    }
    public KillerSudokuType(SudokuType type, ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages, int[][] filledGrid){
        super(type);
        this.setFilledGrid(filledGrid);
        this.cages = cages;
    }


    public KillerSudokuType(SudokuType type, int[][] unfilledGrid, int[][] filledGrid,
                            ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages) {
        super(type, unfilledGrid, filledGrid);
        this.cages = cages;
    }

    public KillerSudokuType(SudokuType sudokuType, int[][] unfilledGrid, int[][] filledGrid) {
        super(sudokuType, unfilledGrid, filledGrid);
    }

    public ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> getCages(){
        return cages;
    }

    public Pair<Integer, List<Pair<Integer, Integer>>> getCage(Pair<Integer, Integer> pos){
        for(Pair<Integer, List<Pair<Integer, Integer>>> cage : cages){
            if(cage.second.contains(pos)){
                return cage;
            }
        }
        return null;
    }
    public ArrayList<Pair<Integer, Integer>> getBox(Pair<Integer, Integer> pos){
        ArrayList<Pair<Integer, Integer>> result = new ArrayList<>();
        for(int row = pos.first/3*3; row < pos.first/3*3+3; row++){
            for(int col = pos.second/3*3; col < pos.second/3*3+3; col++){
                if(!(row == pos.first && col == pos.second)) {
                    result.add(new Pair<>(row, col));
                }
            }
        }
        return result;
    }
    public ArrayList<Pair<Integer, Integer>> getRow(Pair<Integer, Integer> pos){
        ArrayList<Pair<Integer, Integer>> result = new ArrayList<>();
        for(int col = 0; col < 9; col++){
            if(col != pos.second){
                result.add(new Pair<>(pos.first, col));
            }
        }
        return result;
    }
    public ArrayList<Pair<Integer, Integer>> getCol(Pair<Integer, Integer> pos){
        ArrayList<Pair<Integer, Integer>> result = new ArrayList<>();
        for(int row = 0; row < 9; row++){
            if(row != pos.first){
                result.add(new Pair<>(row, pos.second));
            }
        }
        return result;
    }

    public void setCages(ArrayList<Pair<Integer, List<Pair<Integer, Integer>>>> cages){
        this.cages = cages;
    }
    public void addCage(Pair<Integer, List<Pair<Integer, Integer>>> cage){
        this.cages.add(cage);
    }

    public void PrintSudoku() {
        for(int row=0; row < 9; row++){
            if(row % 3 == 0){
                System.out.println(new String(new char[22]).replace("\0", "-"));
            }
            for(int col=0; col < 9; col++){
                if(col % 3 == 0){
                    System.out.print("|");
                }
                if(this.getGrid()[row][col] == 0){
                    System.out.print("  ");
                } else{
                    System.out.print(" " + this.getGrid()[row][col]);
                }
                if(col == 8){
                    System.out.print("|\n");
                }
            }
        }
        System.out.println(new String(new char[22]).replace("\0", "-"));
    }

    public void PrintSudokuStats(){
        super.PrintSudokuStats(this.getGrid());
        //super.PrintSudokuStats();
        if(this.cages == null){
            System.out.println("Cages: null");
            return;
        }
        System.out.println("Num of cages: " + this.cages.size());
        for (Pair<Integer, List<Pair<Integer, Integer>>> cage : this.cages) {
            System.out.println("\nSum: " + cage.first);
            System.out.print("Points: ");
            for (Pair<Integer, Integer> point : cage.second) {
                System.out.print(point.toString() + "   ");
            }
            System.out.print("\n");
        }

    }
}

File: app/src/main/java/com/techatonic/sudokuapp/backend/sudoku/sudokutypes/KillerSudokuType.kt
package com.techatonic.sudokuapp.backend.sudoku.sudokutypes

import kotlin.Pair

class KillerSudokuType : ClassicSudokuType {
    var cages: MutableList<Pair<Int, List<Pair<Int, Int>>>>? = mutableListOf()

    constructor() : super(SudokuType.Killer) {}
    constructor(sudokuType: SudokuType, unfilledGrid: Array<IntArray>, filledGrid: Array<IntArray>) : super(sudokuType, unfilledGrid, filledGrid) {}

    fun getCage(pos: Pair<Int, Int>): Pair<Int, List<Pair<Int, Int>>>? {
        for (cage in cages!!) {
            if (cage.second.contains(pos)) {
                return cage
            }
        }
        return null
    }

    fun getBox(pos: Pair<Int, Int>): ArrayList<Pair<Int, Int>> {
        val result = ArrayList<Pair<Int, Int>>()
        for (row in pos.first / 3 * 3 until pos.first / 3 * 3 + 3) {
            for (col in pos.second / 3 * 3 until pos.second / 3 * 3 + 3) {
                if (!(row == pos.first && col == pos.second)) {
                    result.add(Pair(row, col))
                }
            }
        }
        return result
    }

    fun getRow(pos: Pair<Int, Int>): ArrayList<Pair<Int, Int>> {
        val result = ArrayList<Pair<Int, Int>>()
        for (col in 0..8) {
            if (col != pos.second) {
                result.add(Pair(pos.first, col))
            }
        }
        return result
    }

    fun getCol(pos: Pair<Int, Int>): ArrayList<Pair<Int, Int>> {
        val result = ArrayList<Pair<Int, Int>>()
        for (row in 0..8) {
            if (row != pos.first) {
                result.add(Pair(row, pos.second))
            }
        }
        return result
    }

    fun addCage(cage: Pair<Int, List<Pair<Int, Int>>>) {
        cages!!.add(cage)
    }

    override fun PrintSudoku() {
        for (row in 0..8) {
            if (row % 3 == 0) {
                println(String(CharArray(22)).replace("\u0000", "-"))
            }
            for (col in 0..8) {
                if (col % 3 == 0) {
                    print("|")
                }
                if (grid[row][col] == 0) {
                    print("  ")
                } else {
                    print(" " + grid[row][col])
                }
                if (col == 8) {
                    print("|\n")
                }
            }
        }
        println(String(CharArray(22)).replace("\u0000", "-"))
    }

    override fun PrintSudokuStats() {
        super.PrintSudokuStats(grid)
        //super.PrintSudokuStats();
        if (cages == null) {
            println("Cages: null")
            return
        }
        println("Num of cages: " + cages!!.size)
        for (cage in cages!!) {
            println(
                """
    
    Sum: ${cage.first}
    """.trimIndent()
            )
            print("Points: ")
            for (point in cage.second) {
                print("$point   ")
            }
            print("\n")
        }
    }
}